src/GrumPHP/Process/AsyncProcessRunner.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace GrumPHP\Process;
+
+use GrumPHP\Configuration\GrumPHP;
+use GrumPHP\Runner\TaskResult;
+use Symfony\Component\Process\Process;
+
+/**
+ * Class AsyncProcessRunner
+ *
+ * @package GrumPHP\Process
+ */
+class AsyncProcessRunner
+{
+    /**
+     * @var GrumPHP
+     */
+    private $config;
+
+    /**
+     * @var array
+     */
+    private $processes;
+
+    /**
+     * @var int
+     */
+    private $running;
+
+    /**
+     * AsyncProcessRunner constructor.
+     *
+     * @param GrumPHP $config
+     */
+    public function __construct(GrumPHP $config)
+    {
+        $this->config = $config;
+    }
+
+    /**
+     * @param Process[] $processes
+     */
+    public function run(array $processes)
+    {
+        $this->processes = $processes;
+        $this->running = 0;
+
+        while ($this->watchProcesses()) {
+            usleep(1000);
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    private function watchProcesses()
+    {
+        $shouldRunAgain = false;
+
+        foreach ($this->processes as $key => $process) {
+            $isTerminated = $this->handleProcess($process);
+
+            if ($isTerminated) {
+                unset($this->processes[$key]);
+            }
+
+            $shouldRunAgain = max($shouldRunAgain, !$isTerminated);
+        }
+
+        return $shouldRunAgain;
+    }
+
+    /**
+     * @return bool
+     */
+    private function handleProcess(Process $process)
+    {
+        $isTerminated = false;
+
+        if (!$process->isStarted() && $this->running < $this->config->getProcessAsyncLimit()) {
```

Comment:
Try to avoid else / elseif's.
Just invert the if and return early.
We're trying to folow the [object calisthenics](http://www.slideshare.net/guilhermeblanco/object-calisthenics-applied-to-php) rules.